bigint_func.circom

```diff
@@ -0,0 +1,442 @@
+        if (i < k) {
+            for (var a_idx = 0; a_idx <= i; a_idx++) {
+                prod_val[i] = prod_val[i] + a[a_idx] * b[i - a_idx];
+            }
+        } else {
+            for (var a_idx = i - k + 1; a_idx < k; a_idx++) {
+                prod_val[i] = prod_val[i] + a[a_idx] * b[i - a_idx];
+            }
+        }
+    }
+
+    // now do a bunch of carrying to make sure registers not overflowed. taken from LongToShortNoEndCarry2
+    var out[100]; // length is 2 * k
+
+    var split[100][3]; // first dimension has length 2 * k - 1
+    for (var i = 0; i < 2 * k - 1; i++) {
+        split[i] = SplitThreeFn(prod_val[i], n, n, n);
+    }
+
+    var carry[100]; // length is 2 * k - 1
+    carry[0] = 0;
+    out[0] = split[0][0];
+    if (2 * k - 1 > 1) {
+        var sumAndCarry[2] = SplitFn(split[0][1] + split[1][0], n, n);
+        out[1] = sumAndCarry[0];
+        carry[1] = sumAndCarry[1];
+    }
+    if (2 * k - 1 > 2) {
+        for (var i = 2; i < 2 * k - 1; i++) {
+            var sumAndCarry[2] = SplitFn(split[i][0] + split[i-1][1] + split[i-2][2] + carry[i-1], n, n);
+            out[i] = sumAndCarry[0];
+            carry[i] = sumAndCarry[1];
+        }
+        out[2 * k - 1] = split[2*k-2][1] + split[2*k-3][2] + carry[2*k-2];
+    }
+    return out;
+}
+
+// n bits per register
+// a has k registers
+// p has k registers
+// e has k registers
+// k * n <= 500
+// p is a prime
+// computes a^e mod p
+function mod_exp(n, k, a, p, e) {
+    var eBits[500]; // length is k * n
+    for (var i = 0; i < k; i++) {
+        for (var j = 0; j < n; j++) {
+            eBits[j + n * i] = (e[i] >> j) & 1;
+        }
+    }
+
+    var out[100]; // length is k
+    for (var i = 0; i < 100; i++) {
+        out[i] = 0;
+    }
+    out[0] = 1;
+
+    // repeated squaring
+    for (var i = k * n - 1; i >= 0; i--) {
+        // multiply by a if bit is 0
+        if (eBits[i] == 1) {
+            var temp[200]; // length 2 * k
+            temp = prod(n, k, out, a);
+            var temp2[2][100];
+            temp2 = long_div(n, k, k, temp, p);
+            out = temp2[1];
+        }
+
+        // square, unless we're at the end
+        if (i > 0) {
+            var temp[200]; // length 2 * k
+            temp = prod(n, k, out, out);
+            var temp2[2][100];
+            temp2 = long_div(n, k, k, temp, p);
+            out = temp2[1];
+        }
+
+    }
+    return out;
+}
+
+// n bits per register
+// a has k registers
+// p has k registers
+// k * n <= 500
+// p is a prime
+// if a == 0 mod p, returns 0
+// else computes inv = a^(p-2) mod p
+function mod_inv(n, k, a, p) {
+    var isZero = 1;
+    for (var i = 0; i < k; i++) {
+        if (a[i] != 0) {
+            isZero = 0;
+        }
+    }
+    if (isZero == 1) {
+        var ret[100];
+        for (var i = 0; i < k; i++) {
+            ret[i] = 0;
+        }
+        return ret;
+    }
+
+    var pCopy[100];
+    for (var i = 0; i < 100; i++) {
+        if (i < k) {
```

Comment:
`k` is assumed to be < 100. Document it and check if this assumption can be violated.